### PR TITLE
[GStreamer][WebRTC] Incoming track ID doesn't match sender track ID

### PR DESCRIPTION
**Pull request by philn**
#### 5a644c4173145039030534b9c3d31d7967ab1da8
<pre>
[GStreamer][WebRTC] Incoming track ID doesn&apos;t match sender track ID
<a href="https://bugs.webkit.org/show_bug.cgi?id=258124">https://bugs.webkit.org/show_bug.cgi?id=258124</a>

Reviewed by Xabier Rodriguez-Calvar.

The issues here were that we were handling SDP msid attribute values as ascii instead of unicode
encoded in UTF8 and that we were using our sender track ID as receiver (incoming) track ID.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/webrtc/utf8-sdp.html:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::addTrack):
(WebCore::GStreamerMediaEndpoint::trackIdFromSDPMedia):
(WebCore::GStreamerMediaEndpoint::addRemoteStream):
(WebCore::GStreamerMediaEndpoint::collectTransceivers):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.cpp:
(WebCore::GStreamerPeerConnectionBackend::newRemoteTransceiver):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerPeerConnectionBackend.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerRtpTransceiverBackend.cpp:
(WebCore::GStreamerRtpTransceiverBackend::mid):
* Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp:
(WebCore::RealtimeOutgoingMediaSourceGStreamer::allowedCaps const):
(WebCore::RealtimeOutgoingMediaSourceGStreamer::setSource):

Canonical link: <a href="https://commits.webkit.org/265244@main">https://commits.webkit.org/265244@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/d0b67a8f755bd374516a95f1cbe9f5e3d7a701f9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10717 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11863 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9850 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10229 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12449 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10407 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12788 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10375 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11117 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8611 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12256 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/9015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9252 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16551 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9521 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9404 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12661 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9861 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8002 "1 flakes 1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9020 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9147 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2485 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13268 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9690 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp [hunk 1]
```diff
@@ -847,6 +847,20 @@ MediaStream& GStreamerMediaEndpoint::mediaStreamFromRTCStream(String mediaStream
...
 }
 
+String GStreamerMediaEndpoint::trackIdFromSDPMedia(const GstSDPMedia& media)
+{
+    const char* msidAttribute = gst_sdp_media_get_attribute_val(&media, "msid");
+    if (!msidAttribute)
+        return emptyString();
+
+    GST_LOG_OBJECT(m_pipeline.get(), "SDP media msid attribute value: %s", msidAttribute);
+    auto components = String::fromUTF8(msidAttribute).split(' ');
```
Comment (calvaris) [inline Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:857]: Maybe you can use a StringView here.

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp [hunk 2]
```diff
@@ -861,45 +875,53 @@ void GStreamerMediaEndpoint::addRemoteStream(GstPad* pad)
...
+        return;
     }
 
     if (mediaStreamId.isEmpty()) {
         if (const char* msidAttribute = gst_sdp_media_get_attribute_val(media, "msid")) {
-            auto components = makeString(msidAttribute).split(' ');
-            if (components.size() == 2)
-                mediaStreamId = components[0];
+            GST_LOG_OBJECT(m_pipeline.get(), "SDP media msid attribute value: %s", msidAttribute);
+            auto components = String::fromUTF8(msidAttribute).split(' ');
```
Comment (calvaris) [inline Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:904]: Maybe this can be a StringView.

**Review by calvaris — APPROVED** [review]

File: Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp (hunk 1 above)
Comment (philn) [reply Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:857]: It has no `fromUTF8()`, only `fromLatin1()`. It's unclear if StringView can handle UTF8.

Comment (philn) [reply Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:857]: Maybe we can micro-optimize this once we've made significant progress on the webrtc backend. Landing as-is.

**Pull request merged**